### ISSUE TITLE
Use filter with a start block from the end of the previous subscription when resubscribing

### ISSUE THREAD

**Issue by yondonfu** [livepeer/go-livepeer#315]
There is an an edge case with subscription retries where a transcoder's new job subscription could end due to a closed websocket connection with Infura, a new job is created that is assigned to the transcoder and when the transcoder resubscribes to new jobs it will have missed the new job that was just created. 

I think we should be able to add some logic to track the block at which a subscription ends and then update the filter used for the subsequent resubscribe so that it will filter for events starting from the block at which the previous subscription ended so that the transcoder would still receive the new job event that was fired during the brief period of time that it wasn't subscribed. Similar logic could be used for other events as well.

Additionally, if subscriptions keep failing we should probably just stop resubscribing after a number of retries in the case where there is an issue with Infura such that a websocket connection just cannot be opened successfully. We could inform the user of such a case

**Comment by ericxtang** [comment]
From @yondonfu - 

> Perhaps an extension to this proposal that is related to #315 is that a transcoder keeps track of the last block that it was assigned a job and we introduce an option in the CLI, something like "Sync Jobs". When that option is triggered, the transcoder can filter for job events from the last block that it was assigned a job until the current block and check if it was assigned to any of them and start doing work for the job.

**Comment by j0sh** [comment]
> if subscriptions keep failing we should probably just stop resubscribing after a number of retries

For long-lived nodes such as a transcoder, it is probably better (and harmless) to keep re-trying for as long as needed, maybe with some backoff up to a few minutes. We certainly should still inform the user though.

**Comment by dob** [comment]
Related, but perhaps not exactly the same, is "subscribing" to the actual streams on the network. When a transcoder restarts does it need to resubscribe to any stream that has not been explicitly ended for which it was assigned? This is also realted to the concept of long running reserved jobs, where a broadcaster gets assigned a transcoder, and may start/stop streaming intermittently for hours or days until the endBlock of the job.

Just because it sends the subscribe request back towards the broadcaster doesn't mean that it necessarily will continue to get segments, if the broadcaster is in fact, not broadcasting. But it won't continue to get any segments it should get if it doesn't resubscribe. (I also may be off on the updated architecture)

**Comment by j0sh** [comment]
Posted some related thoughts on re-subscription here  https://github.com/livepeer/go-livepeer-basicnet/issues/21#issuecomment-372748851 , related to the work for https://github.com/livepeer/go-livepeer-basicnet/issues/29

**Comment by yondonfu** [comment]
Apparently according to this open issue https://github.com/ethereum/go-ethereum/issues/15063 Geth also does not respect the `fromBlock` and `toBlock` parameters for log subscriptions. I've also verified this locally - if you create a log subscription with a set of topics and a `fromBlock` value, the `fromBlock` value is ignored and the subscription will only return new logs starting from the current block. Thus this approach does not currently work - perhaps more reason to move to the architecture presented in #492 

EDIT: I also tested this with Parity and it appears that the `fromBlock` parameter is also not respected for the `eth_subscribe` RPC call for logs. I tried subscribing to logs on Kovan with a `fromBlock` in the past using the RPC call `{"method":"eth_subscribe","params":["logs", {"fromBlock":"0x7C7EF8"}],"id":1,"jsonrpc":"2.0”}` both using Infura and a local Parity node. In both cases, I would only start receiving logs that are included in new blocks.
